src/test/java/com/example/demo/controller/BooksControllerTest.java

```diff
@@ -60,4 +61,14 @@ public void searchTest() throws Exception {
                         .isOk())
                 .andExpect(content().string(containsString("1")));
     }
+
+    @Test
+    public void filterTest() throws Exception {
+        when(service.findBooksWithParams(any())).thenReturn(result);
+        mockMvc.perform(get("/filter"))
+                .andDo(print())
+                .andExpect(status()
+                        .isOk())
+                .andExpect(content().string(containsString("1")));
```

Comment:
tutaj dostajesz jakiś JSON z danymi. fajnie byłoby sprawdzić czy ten JSON jest poprawny.  

1. czy ma poprawną ilość obiektów w liście
2. czy lista ma dobrą wielkość
3. czy obiekt jest uzupełniony w polu w którym chcesz aby był uzupełniony.

Przydatnym mechanizmem jest tutaj `jsonPath`

przykład: `andExpect(jsonPath("$[0]").exists())` sprawdza, czy pierwszy obiekt w liście istnieje. Poczytaj sobie o tym i spróbuj poprawić ten test aby sprawdzał czy pole które ustawiłaś jest poprawnie uzupełnione. Teraz jeśli w jakimkolwiek miejscu będziesz miała `1` to ten test przejdzie.

Reply:
done